### DISCUSSION TITLE
Questions around train_test_split_edges and negative_sampling functions.

### DISCUSSION THREAD

**Discussion by sbonner0** [pyg-team/pytorch_geometric#2655] [Q&A] [answered]
Hey I had a few questions/thoughts on the `train_test_split_edges` and `negative_sampling` functions and their use in the context of link prediction. 

- I noticed that during the process of the train_test_split_edges, that the following happens: `to_undirected(data.train_pos_edge_index) `. I had thought that the edge index tensor needed to explicitly contain the train edges in both directions, as mentioned on the intro page? If it were creating my own train/test split not using this function, would I need to ensure the graph was undirected as well? 
- I guess this is an issue with all negative sampling approaches, but there is nothing stopping the positive test/val edges being sampled at negatives when this function is used repeatedly - for example, in the forward pass of the GAE models. There is also the issue of same negatives randomly being used in both training and testing. Is there anything that can be done to mitigate this?

**Comment by rusty1s** [comment, marked as answer]
1. Yes, currently `train_test_split_edges` assumes an undirected graph represented by `edge_index`. During splitting, we then make sure that both edges are contained in the same split. If you simply want to split your directed edges, I guess you can simple split `edge_index` based on a random permutation:
```python
perm = torch.randperm(edge_index.size(1))
pos_train_edge_index = edge_index[:, perm[:1000]]
...
```
2. Yes, this is indeed the case. It might well happen that positive validation edges and test edges appear as negative samples during training. We are not allowed to acknowledge their existence during training to prevent any data leakage. Although this seems to be counter-intuitive at first glance, it is to be expected that such a noise in the learning signal is averaged out during optimization. In theory, you can sample negative edges based on positive train/val/test edges, but note that this will leak validation and test data to your training procedure.

**Reply by sbonner0** [reply]
Thanks @rusty1s - very clear explanations!
